### DISCUSSION TITLE
Layout using parallel routes throw TypeScript invalid "default" export error on next build

### DISCUSSION THREAD

**Discussion by csalmeida** [vercel/next.js#58275] [Help] [answered]
### Summary

Hi everyone, I have a layout that makes use of the Parallel Routes feature.
It works great when running `next dev` but when running `next build`, it throws the following TypeScript error:

```bash
Type error: Layout "app/archives/layout.tsx" has an invalid "default" export:
  Type "Props" is not valid.
```

This is how the layout is setup, any ideas on what might be causing this? 

```TypeScript
type Props = {
  children: React.ReactNode
  archivePanel: React.ReactNode
  archiveForm: React.ReactNode
}

export default function Layout({
    children,
    archivePanel,
    archiveForm
}: Props) {
    return (
        <>
            {children}
            {archivePanel}
            {archiveForm}
        </>
    )
}
```

Any thoughts would be really helpful 🙏

### Additional information

```js
App is using Next `14.0.1`.

This snippet uses parallel routes: https://nextjs.org/docs/app/building-your-application/routing/parallel-routes

The error is thrown at `next build`:


Type error: Layout "app/archives/layout.tsx" has an invalid "default" export:
  Type "Props" is not valid.
```

### Example

[Example Project](https://github.com/csalmeida/nextjs-layout-typescript-error)

**Comment by icyJoseph** [comment]
I guess the issue is, `archivePanel` vs `archive`?

```ts
type Props = {
  children: React.ReactNode
  archive: React.ReactNode // This one looks wrong? should it `archivePanel`?
  archiveForm: React.ReactNode
}
```

Or maybe the other way around, the `archivePanel`, prop passed to Layout is incorrect?

**Reply by csalmeida** [reply]
Thanks @icyJoseph, that's a good shout! 

That is wrong in the snippet but I don't have that mistake in the original project, I just edited by accident and didn't notice, my apologies.

Given that the layout actually works as expected and described on the Next.js docs, a way to at least ignore the TS check would be good but `@ts-ignore` is not doing the trick here unfortunately. Was wondering if anyone had maybe come across this before. 

The correct snippet looks like this:

```typescript
type Props = {
  children: React.ReactNode
  archivePanel: React.ReactNode
  archiveForm: React.ReactNode
}

export default function Layout({
    children,
    archivePanel,
    archiveForm
}: Props) {
    return (
        <>
            {children}
            {archivePanel}
            {archiveForm}
        </>
    )
}
```

**Reply by icyJoseph** [reply]
Could you try to setup a small reproduction demo? This is starting to sound like a bug (but NOT one yet)... it only happens with nested layouts right?

**Reply by csalmeida** [reply]
The layout is nested in `/app/archives`.
Absolutely, I managed to replicate it here: https://github.com/csalmeida/nextjs-layout-typescript-error

After setting up the project, run `npm run build` and the output should be:

```bash
> nextjs-layout-typescript-error@0.1.0 build
> next build

   ▲ Next.js 14.0.2

 ✓ Creating an optimized production build
 ✓ Compiled successfully
   Linting and checking validity of types  ...Failed to compile.

app/(auth)/archives/layout.tsx
Type error: Layout "app/(auth)/archives/layout.tsx" has an invalid "default" export:
  Type "Props" is not valid.
```

Appreciate the support 🙏

**Reply by tholder** [reply, marked as answer]
This is a really stupid way to solve this, you can't even ts ignore the entire file but if it has the default expected type it passes build, so this works:

```
type LayoutProps = {
    children?: React.ReactNode
}

type LayoutPropsExtended = {
    children?: React.ReactNode
    archivePanel?: React.ReactNode
    archiveForm?: React.ReactNode
}

export default function Layout(props: LayoutProps | LayoutPropsExtended) {
    const { children, archivePanel, archiveForm } = {
        ...props,
        archiveForm: undefined,
        archivePanel: undefined,
    }

    return (
        <div className="h-full">
            {children}
            {archiveForm}
            {archivePanel}
        </div>
    )
}

**Reply by csalmeida** [reply]
It did work; I just tested it and the build passes okay so it could be a workaround for now. 🎉

![CleanShot 2023-11-18 at 00 18 22@2x](https://github.com/vercel/next.js/assets/12223613/456e569e-d99f-4b66-9355-e3c1d725d149)

**Reply by artemis-prime** [reply]
I'm curious though, OP... You're passing 3 `ReactNode`'s into a layout.  How are you doing this?  I thought layouts only nest via their `children` prop.  Are you using cloneElement in the parent to pass the props?  I ask because I'd love to have layouts with multiple slots working w ssr!  In a client only environment, you just use renderprops or pass in ReactNodes, but I didn't think that was possible to a layout, which is a server component.

**Reply by csalmeida** [reply]
It's been a while since I looked at this project but to make this happen I used a feature called [Parallel Routes](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes).

In my example, [you can see that `archiveForm` and `archivePanel` are marked as parallel routes here](https://github.com/csalmeida/nextjs-layout-typescript-error/tree/main/app/(auth)), then [these pages can be accessed in the `/archives/layout.tsx` file](https://github.com/csalmeida/nextjs-layout-typescript-error/blob/main/app/(auth)/archives/layout.tsx).

I don't remember all the heuristics but I believe as long as you prefix the page directory with `@`, and reference it in your layout, NextJS will do some magic to tie everything together.

**Comment by thisisankur0** [comment]
For anyone still facing this issue: Build error only comes up if you have declared the page inside props but the file doesn't exist inside the folder layout.

**Reply by onatyy** [reply]
What do you exactly mean by this. For example I'm getting the error: Type error: Page "app/integrations/page.tsx" has an invalid "default" export:
  Type "HTMLAttributes<HTMLDivElement> & RefAttributes<HTMLDivElement>" is not valid.

**Reply by artemis-prime** [reply]
Please clarify your answer?  I can't parse what you're saying.

**Reply by charltoons** [reply]
Not sure if this is exactly what @thisisankur0 was experiencing, but for me I had copied my `layout.tsx` from somewhere else and forgot to move the parallel route props.

example:
```tsx
export default async function MyLayout(props: { 
  children: ReactNode, 
  header: ReactNode // I had `@header/...` parallel route files for another layout but not one for this layout. This prop was unused anyway, so removing it here fixed my issue.
}) {
```

**Reply by thisisankur0** [reply]
Yep exactly the same as @charltoons, thanks for a more clarifying reply.

**Reply by priorArthur** [reply]
Thx for this.  Finally an explanation that makes sense.  Wish the error logs just said what your comment says.  Maybe they can put a link to the comment ;-)

**Comment by Cyclodex** [comment]
I had the same issue and came across this post.
However for me it was super strange, as it worked locally , but in the docker setup it did not - and failed with the same error as mentioned in this issue.

I found out that:
VSCode was using typescript @ `5.3.2` OK _(it was not intended to use the vscode version, I usually use the package.json one)_
Docker build did use typescript @ `5.3.3` NOT OK _(which comes from the package-lock.json version)_

Now I increased the version to typescript "latest" @ `5.4.3` OK
Interestingly the issue is gone.
Maybe this helps anyone...

**Reply by uhrohraggy** [reply]
FYI I'm dealing with this during `next build` in `typescript`@ `5.5.4` unfortunately for a nested `layout.tsx` that has dynamic path segments (i.e. having a `[foo_key]/layout.tsx` using Next.js 14.2.7 with App Router). Works perfectly in dev but fails during build

**Reply by uhrohraggy** [reply]
Ah, PEBKAC, but I was mixing Next.js 12 style with newer App router approach.

so a client component file like `[lang]/client/[foo_key]/layout.tsx`, change:

```
type FooKeyLayoutProps = {
  children: ReactNode;
  params: {
    lang: Locale;
    foo_key: string;
  };
  searchParams?: {}
};

const Layout = async ({ children, params }: FooKeyLayoutProps) => {
  const { lang, foo_key } = params
  ```
  
  to
  
```
import { useParams } from 'next/navigation'
  
type FooKeyLayoutProps = {
  children: ReactNode;
};

const Layout = async ({ children }: FooKeyLayoutProps) => {
  // Vars
  const {lang, foo_key} = useParams()
  ```

**Reply by Jamiewarb** [reply]
Mine was similar, in case anyone finds this useful.

I had the props incorrect, and wasn't awaiting params

```
// incorrect
export default async function Layout({
  children,
  siteId,
}: LayoutProps) { ... };

// correct
export default async function({
  children,
  params,
}: LayoutProps) {
  const { siteId } = await params;
  ...
}
```

**Comment by swayok** [comment]
I have the same problem. And the worst is that it does not indicated in IDE, nor it is present in dev mode. Also it happens only on top-level layout and does not happen on other layouts or pages. And it works perfectly fine... Well, that is a way too magic to understand.
My solution was to cast the export type:
```
export interface LocalizationProps {
    locale: LocaleCode
    localeConfig: LocaleConfig
    trans: Translations
}

export interface LayoutProps {
    children?: React.ReactNode
}

export interface PageProps {
    params: Promise<{
        locale: LocaleCode
    }>
    searchParams: Promise<AnyObject<string>>
}

export interface LocalizedLayoutProps extends LayoutProps, PageProps, LocalizationProps {
}

async function LocaleLayout(props: LocalizedLayoutProps) {

}

// Here we cast export to React.FunctionComponent<LayoutProps> to shut up that annoying error.
export default withLocalizationProps<LocalizedLayoutProps>(LocaleLayout) as React.FunctionComponent<LayoutProps>
```
